File: Code/Python/Composite/composite.py
class Structure:
    def getChild(self, i):
        raise NotImplementedError
    def add(self, child):
        raise NotImplementedError
    def remove(self, child):
        raise NotImplementedError

class Railroad(Structure):
    def __init__(self, length):
        self.length = length
    def getMaintananceCost(self):
        return 2*self.length

class Tunnel(Structure):
    def __init__(self, length):
        self.length = length
    def getMaintananceCost(self):
        return 5*self.length

class Station(Structure):
    def getMaintananceCost(self):
        return 20

class Network:
    def __init__(self):
        self.children = []

    def getChild(self, i):
        return self.children[i]
    def add(self, child):
        self.children.append(child)
    def remove(self, child):
        self.children.remove(child)
    def getMaintananceCost(self):
        return sum(child.getMaintananceCost() for child in self.children)

tunnelNetwork = Network()
tunnelNetwork.add(Tunnel(5))
tunnelNetwork.add(Railroad(5))

stationNetwork = Network()
stationNetwork.add(Station())
stationNetwork.add(Railroad(10))

bigNetwork = Network()
bigNetwork.add(tunnelNetwork)
bigNetwork.add(stationNetwork)
print(bigNetwork.getMaintananceCost())

File: Code/CSharp/Composite/composite.cs
using System.Collections.Generic;
using System.Linq;

interface NetworkElement {
    NetworkElement GetChild(int i);
    void Add(NetworkElement child);
    void Remove(NetworkElement child);
    int GetMaintananceCost();
}

abstract class Structure : NetworkElement {
    public NetworkElement GetChild(int i){
        throw new System.NotImplementedException();
    }
    public void Add(NetworkElement child) {
        throw new System.NotImplementedException();
    }
    public void Remove(NetworkElement child) {
        throw new System.NotImplementedException();
    }

    public abstract int GetMaintananceCost();
}


class Network : NetworkElement {
    private List<NetworkElement> children = new List<NetworkElement>();

    public NetworkElement GetChild(int i) {
        return children[i];
    }
    public void Add(NetworkElement child) {
        children.Add(child);
    }
    public void Remove(NetworkElement child) {
        children.Remove(child);
    }
    public int GetMaintananceCost() {
        return children.Sum(child => child.GetMaintananceCost());
    }
}

class Railroad : Structure {
    private int length;
    public Railroad(int length) {
        this.length = length;
    }

    public override int GetMaintananceCost() {
        return length*2;
    }
}

class Tunnel : Structure {
    private int length;
    public Tunnel(int length) {
        this.length = length;
    }
    public override int GetMaintananceCost() {
        return this.length*5;
    }
}

class Station : Structure {
    public override int GetMaintananceCost() {
        return 20;
    }
}

class Tester {
    static void Main(string[] args) {
        NetworkElement tunnelNetwork = new Network();
        tunnelNetwork.Add(new Tunnel(5));
        tunnelNetwork.Add(new Railroad(5));

        NetworkElement stationNetwork = new Network();
        stationNetwork.Add(new Station());
        stationNetwork.Add(new Railroad(10));

        NetworkElement bigNetwork = new Network();
        bigNetwork.Add(tunnelNetwork);
        bigNetwork.Add(stationNetwork);

        System.Console.WriteLine(bigNetwork.GetMaintananceCost());
    }
}
